Supervisor model

tsseppa opened the pull request:
### Version 1.0

Supervisor was created to ´src/enchanted_surrogates/supervisor/supervisor.py´

General architecture:
- Some tasks previously in ´run.py´ was moved to ´supervisor.py´ (path creation, executor intializing)
- Some tasks of ´Executor´ class was moved to ´supervisor.py´, see changes in ´base_executor.py´
- Sampler is no more initialized by executor. Now sampler is initialized by supervisor, which gives samples and sampler to executor (due to sampler having the function ´register_futures´ which should be accessed during execute). 
- ´local_executor.py´ was modified to work with new supervisor architecture
- Supervisor prompts user a question if ´base_run_dir´ has content inside and user wants to delete content. Content should be in current implementation deleted, as the ´enchanted_datapoints.csv´ would otherwise be appended, which breaks thing. This is due to decision to name folders with ´d{nesting_depth}_b{batch_number}_r{run_number/sample_number}´, which could possibly help with the issue Many Small Files Issue. Can be changed back if original implementation is preferred.

### Nesting:

- Supervisor now has 1...n executors and samplers:
  - Config files now have separate dicts for executors, samplers and runners. Supervisor key contains a list of run groups, where it is specified which executor, sampler and runner are used together. By adding more elements to the list, elements are nested each other.
  - All executors, samplers and runners have a name in the config files, and those names can be used as keys to reuse them in run groups.
  - Supervisor loops over all executors, runs them one by one and supplies the results of previous run as input data to the next run.
- Separate DaskNestedExecutor has been removed.


### Configuration
Some modifications made for configuration:
 - Supervisor added to config.yaml hierarchy
 - `base_run_dir` moved from executor to supervisor in config hierarchy
 - supervisor can have configuration argument `summary_datatype` with type `parquet`. This sets summary to be parquet instead of csv. ´pyarrow´ was added to dependencies
 - run_order added to supervisor, see file changes

#163 hdf5 storage also part of pull request. It has storage config attribute.

**Some changes will come from #170 **

Commit:
feat: initial creation for supervisor

Commit:
Added pyarrow dependency

Commit:
change: moved functionality to supervisor

Commit:
change: moved functionality to supervisor

Commit:
feat: created supervisor

Commit:
change: modified to fit supervisor implementation

Commit:
change: moved functionalities to supervisor

Commit:
feat: created supervisor

Commit:
feat: improved initial supervisor

Commit:
feat: initial supervisor creation

Commit:
fix: function returns actual value

Commit:
refactor: modified to fit supervisor module

Commit:
refactor: modified to fit supervisor architecture

Commit:
bug: removed unnecessary line

Commit:
refactor: modified to fit supervisor

Commit:
feat: initial supervisor creation

Commit:
docs: Created docstrings

Commit:
fix: pylint warnings in supervisor module

Commit:
docs: appended docstring to match current architecture

Commit:
fix: fix pylint warnings in run.py

JaaskelainenL commented on src/enchanted_surrogates/supervisor/supervisor.py:
```diff
@@ -0,0 +1,197 @@
+                value = input(
+                    str(os.path.abspath(base_run_dir))
+                    + "\nFolders have content. Do you want to delete data in existing folders? y/N "
+                )
```
btw this doesn't work when running with sbatch

tsseppa replied:
i guess there is still cases where when running with sbatch there can be something in base_run_dir, so we have to find a way to 1) clean it in code OR 2) give instructions to user to delete it?

JaaskelainenL replied:
running with sbatch does result in an output in base_run_dir and it is not empty. i think checking if stdout is a tty is a good start. then you have to make a choice to either: a) just delete it or b) just quit and prompt the user to delete it manually

JaaskelainenL replied:
resolved this

vanhalaj committed:
refactor: Joblib executor now works with supervisor

vanhalaj committed:
test: fixed unit test for joblib executor

vanhalaj committed:
test: fixed unit test for local executor

vanhalaj committed:
test: Renamed test functions in joblib and local executor unit tests

vanhalaj commented:
Refactored joblib executor to work with supervisor module and fixed unit tests for local and joblib executor. Dask executor unit tests still need to be updated.

vanhalaj committed:
test: Added unit tests for supervisor module

JaaskelainenL committed:
Feature/enc 121 works for dask executor (#160)

* Start working on migrating dask_executor

* fix samples

JaaskelainenL commented:
dask nested executor should be on its own branch since it might require a lot of changes

JaaskelainenL committed:
Fix issue when not a tty

vanhalaj committed:
test: Fixed unit tests of dask executor

vanhalaj committed:
fix: Removed isinstance from dask executor

vanhalaj committed:
fix: Removed old code that created dataset.csv to the wrong place

vanhalaj committed:
refactor: Removed config file reading from executors

vanhalaj committed:
feat: Parsing of nested config files added to supervisor

vanhalaj committed:
chore: Added config file found from nested executor workflow test

vanhalaj committed:
chore: Added config file that uses the new style of nesting

vanhalaj committed:
feat: Until better implementation, added function to give runner config to executor after initialization

vanhalaj committed:
feat: Child samplers are no longer imported in NestedSampler, instead list of Sampler objects is given in constructor params

vanhalaj committed:
feat: Supervisor now works for nested samplers, but only with one executor

vanhalaj committed:
chore: Added updated example config files utilizing nested sampling, these are only for temporary testing

tsseppa committed:
Hdf5 storage format (#163)

* feature: initialize hdf5 configuration

* feature: adjusted hdf5 configuration

* feature: initial hdf5 creation added

* feature: added h5py to dependency

* docs: documentation for create hdf5

* test: created test for creating hdf5

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* fix: Disable non-member linting in supervisor test

---------

Co-authored-by: Touko Seppä <touko.seppa@aalto.fi>
Co-authored-by: Samuel Hughes <hughes.posti@gmail.com>

tsseppa committed:
Hdf5 storage format (#163)

* feature: initialize hdf5 configuration

* feature: adjusted hdf5 configuration

* feature: initial hdf5 creation added

* feature: added h5py to dependency

* docs: documentation for create hdf5

* test: created test for creating hdf5

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* fix: Disable non-member linting in supervisor test

---------

Co-authored-by: Touko Seppä <touko.seppa@aalto.fi>
Co-authored-by: Samuel Hughes <hughes.posti@gmail.com>

tsseppa committed:
Hdf5 storage format (#163)

* feature: initialize hdf5 configuration

* feature: adjusted hdf5 configuration

* feature: initial hdf5 creation added

* feature: added h5py to dependency

* docs: documentation for create hdf5

* test: created test for creating hdf5

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* test: fixed test

* fix: Disable non-member linting in supervisor test

---------

Co-authored-by: Touko Seppä <touko.seppa@aalto.fi>
Co-authored-by: Samuel Hughes <hughes.posti@gmail.com>

JaaskelainenL committed:
Fix: Local tests on Dask executor

Now checks are run if the run is on SLURM or not to prevent issues from
occuring.

vanhalaj committed:
feat: New nesting logic that works similarly as DaskNestedExecutor

vanhalaj committed:
refactor: remove unused code

vanhalaj committed:
test: fixed unit tests for Supervisor

vanhalaj committed:
fix: Filter is now optional in wait_all_processes and all_processes_done

vanhalaj committed:
fix: Restored nested sampler

vanhalaj committed:
fix: Removed default value in nested sampler

vanhalaj committed:
refactor: Moved nested imports of samplers and executors out of supervisor class

vanhalaj committed:
tests: Refactored configs/setups in unit tests

vanhalaj committed:
fix: Changed runner_config to be optional constructor argument of executors

vanhalaj committed:
refactor: removed unused code

JaaskelainenL committed:
fix: Throw exception if base_run_dir is not set

JaaskelainenL committed:
fix: Throw exception if base_run_dir is not set

JaaskelainenL committed:
fix: Throw exception if base_run_dir is not set

JaaskelainenL committed:
fix: Throw exception if base_run_dir is not set

vanhalaj committed:
fix: Filter run dir names for both depth and batch number

vanhalaj committed:
chore: Updated example config files for nested executors

vanhalaj committed:
feat: Removed dask nested executor

vanhalaj committed:
test: Fixed hdf5 unit tests to work with multiple supervisors, samplers and runners

vanhalaj committed:
test: Fixed hdf5 unit tests to work with multiple supervisors, samplers and runners

vanhalaj committed:
tests: Fix hdf5 unit test failing

vanhalaj committed:
fix: Fixed how class name is got for hdf5 attributes

JaaskelainenL committed:
Fix: Crashing on Windows (ps)

Commit:
fix: runs without storage parameter defined

Commit:
docs: appended class docstring to include create_hdf5

vanhalaj committed:
test: Redone automated-no-HPC workflow tests to work with supervisor and new config style

Commit:
docs: supervisor.md creation

vanhalaj committed:
test: Workflow tests fixed for simple and nested workflows

vanhalaj committed:
chore: Updated dask local example config

vanhalaj committed:
test: LUMI workflow test uses supervisor

vanhalaj committed:
doc: Updated docstring for 'filter' arguments

vanhalaj committed:
doc: Added explanation comment

vanhalaj committed:
chore: Updated remaining example configs

vanhalaj commented on docs/index.md:
````diff
@@ -75,7 +75,7 @@ python enchanted-surrogates/src/run.py -cf path/to/config/file
+The configuration file should be in YAML format and specify the runner, sampler, executor, supervisor and other parameters needed for the simulation. 
   
 ```yaml
 executor:
````
Add run_order to supervisor and executors, samplers, runners

vanhalaj commented on docs/index.md:
```diff
@@ -107,15 +112,19 @@ Modules for generating samples according to some rule set.
+The supervisor is the entry point. The supervisor reads parameters and has Sampler, Executor and Runner. 
+Supervisor initializes Sampler and fetches samples from it. 
+Supervisor gives samples to executor. Executor initializes some cluster or job queue or similar. 
+A sample is sent to Runner. Runner or Parser tied to it creates files. Supervisor creates data structures 
```
Executor calls run simulation task and sends sample to runner

vanhalaj commented on docs/supervisor.md:
````diff
@@ -0,0 +1,45 @@
+
+```yaml
+supervisor:
+    base_run_dir: "path/to/folder"
````
Add run_order

vanhalaj commented on src/enchanted_surrogates/executors/dask_executor.py:
```diff
@@ -49,20 +55,23 @@ def __init__(self, *args, **kwargs):
+        # self.type = kwargs.get('type')
+        # if self.sampler_config:
+        #     self.sampler_type = self.sampler_config.pop("type")
+        #     self.sampler = import_sampler(
```
Remove unused code

Commit:
docs: fixed documentation

Commit:
docs: added run_order to example

vanhalaj committed:
chore: Fixed linter issues for supervisor module

vanhalaj committed:
Merge branch 'feature/enc-89-create-a-supervisor-module' of github.com:DIGIfusion/enchanted-surrogates into feature/enc-89-create-a-supervisor-module

Commit:
refactor: deleted unused code

awlishar reviewed (commented):
Tested and works well. Amazing job!!
I will do a couple other tests tomorrow so don't merge just yet.

awlishar commented:
Approved for merge. :)
